app/controllers/comments_controller.rb

```diff
@@ -381,6 +412,18 @@ def reject
     redirect_to_all_comments(@comment.ultimate_parent, show_comments: true)
   end
 
+  # PUT /comments/1/freeze
+  def freeze
+    if !@comment.on_ice?
+      @comment.full_set.each { |c| c.mark_frozen! }
```

Comment:
FYI, I believe that [AO3-5939](https://otwarchive.atlassian.net/browse/AO3-5939) means we don't know whether `full_set` will return the right set of comments. It should in most cases, because AO3-5939 is a concurrency issue, but there may be cases where it doesn't.

(If AO3-5939 does happen, it'll affect random comments, but permanently, not intermittently. Because if the race condition causes the `threaded_left` and `threaded_right` values to be incorrectly computed while the new comment is being inserted, those incorrect values will be saved to the database. And that means that strange things will always happen when trying to freeze/unfreeze those comments, and likely other comments nearby.)

Reply:
I wonder if that explains the issue I ran into with [this test](https://github.com/otwcode/otwarchive/pull/3893/files#diff-c37d231b9eefa56b4c32761eb6bf136bR710), where the comments had to be created in a specific order in the specs, which would presumably be creating the comments so quickly it might as well be simultaneous.

I guess the way to avoid using `full_set` would be to:

1. Get the tag, admin post, or chapter of the comment we are freezing.
2. Get all the comments on that tag, admin post, or chapter.
3. Look through that set of comments to find any comments whose commendable is the comment we are freezing.
3. Then look through our set of comments _again_ to find any with _those_ as commentables.
4. And so on, until we don't find any more comments.
5. Finally, freeze all the comments we've found.

...however, I just looked, and `find_all_comments` relies on `all_children`, which also relies on `threaded_left` and `threaded_right`. 🤦  But would that perhaps still work with AO3-5939? It doesn't sound like the comments become totally detached and unfindable -- just ordered incorrectly.

Reply:
I wouldn't be surprised if it came up during tests. It should be possible to simulate the issue by creating a collection of nested comments, and deliberately not reloading the parent comment before adding its children. Which is precisely the sort of situation that you could end up with when crafting comments by hand for a test.

Unfortunately, it's not exclusively an issue of order. Consider a comment tree like the following:
```
Comment A (threaded_left: 1, threaded_right: 6)
- Comment B (threaded_left: 2, threaded_right: 3)
- Comment C (threaded_left: 4, threaded_right: 5)
```
Now suppose that two comments are added in quick succession: Comment D, a reply to Comment B; and Comment E, a reply to Comment C. Comment D is inserted at index 3 (the `threaded_right` value of Comment B), so it'll increment all higher `threaded_left`/`threaded_right` values by 2, so that the tree of comments becomes:
```
Comment A (threaded_left: 1, threaded_right: 8)
- Comment B (threaded_left: 2, threaded_right: 5)
- - Comment D (threaded_left: 3, threaded_right: 4)
- Comment C (threaded_left: 6, threaded_right: 7)
```
So far, everything is correct. Now Comment E is inserted. Because it's lagging a little behind the database, it thinks it's being inserted at index 5, the `threaded_right` value of Comment C before Comment D was inserted and changed everything. So it'll bump up any `threaded_left`/`threaded_right` values that are at least 5, resulting in the following set of comments:
```
Comment A (threaded_left: 1, threaded_right: 10)
- Comment B (threaded_left: 2, threaded_right: 7)
- - Comment D (threaded_left: 3, threaded_right: 4)
- Comment C (threaded_left: 8, threaded_right: 9)
- - Comment E (threaded_left: 5, threaded_right: 6)
```
Comment E is nested under Comment C according to its `commentable_id`/`commentable_type` values, but its `threaded_left`/`threaded_right` values are more appropriate for a comment nested under Comment B. And it'll be included in Comment B's `full_set` and `all_children`, not in Comment C's. The same will be true of any descendants of Comment E that might be added in the future.

(I think it may also simultaneously mangle the `threaded_right` of Comment C, setting it to 7, leaving it with a `threaded_left` larger than its `threaded_right`. But I haven't actually checked.)

Reply:
Thanks for that explanation! I'll link to this comment on AO3-5939, since it may be helpful there.

>Comment E is nested under Comment C according to its commentable_id/commentable_type values, but its threaded_left/threaded_right values are more appropriate for a comment nested under Comment B. And it'll be included in Comment B's full_set and all_children, not in Comment C's. The same will be true of any descendants of Comment E that might be added in the future.

That sounds fine for my purposes. I'm only interested in the commentable columns and the ultimate parent (or pseudo ultimate parent, if I can swing it: getting all the comments for a chapter would be preferable to getting all the comments for the work as a whole). As long as Comment E shows up threaded under _some_ comment on the ultimate parent, it should be fine. 

(And indeed, the test scenario that failed with the `full_set` approach works with this new one.)

Reply:
As long as you don't use anything that filters based on `threaded_left`/`threaded_right` values, I believe you should be okay.

Alternatively, I could just try to fix AO3-5939.

Reply:
Excellent. 

I'm happy to do it either way, so if you get a chance to fix AO3-5939, that would be great! (No rush, though, since we're doing Rails next, and that will take a bit of testing time.) If not, I'll just leave a comment in my code saying it can be changed once AO3-5939 is fixed.